Need to drop `MAX_RETURN` definition as `list_group_images()` isn't there anymore:
```
FAILED: src/librbd/CMakeFiles/rbd_internal.dir/mirror/GroupPromoteRequest.cc.o 
/usr/local/bin/sccache /usr/bin/clang++-19 -DBOOST_ASIO_DISABLE_THREAD_KEYWORD_EXTENSION -DBOOST_ASIO_HAS_IO_URING -DBOOST_ASIO_NO_TS_EXECUTORS -DBOOST_DATE_TIME_DYN_LINK -DBOOST_DATE_TIME_NO_LIB -DBOOST_IOSTREAMS_DYN_LINK -DBOOST_IOSTREAMS_NO_LIB -DBOOST_PROGRAM_OPTIONS_DYN_LINK -DBOOST_PROGRAM_OPTIONS_NO_LIB -DBOOST_RANDOM_DYN_LINK -DBOOST_RANDOM_NO_LIB -DBOOST_REGEX_DYN_LINK -DBOOST_REGEX_NO_LIB -DBOOST_SYSTEM_DYN_LINK -DBOOST_SYSTEM_NO_LIB -DBOOST_THREAD_DYN_LINK -DBOOST_THREAD_NO_LIB -DHAVE_CONFIG_H -D_FILE_OFFSET_BITS=64 -D_GNU_SOURCE -D_REENTRANT -D_THREAD_SAFE -D__CEPH__ -D__STDC_FORMAT_MACROS -D__linux__ -I/ceph/build/src/include -I/ceph/src -isystem /opt/ceph/include -isystem /ceph/build/include -isystem /ceph/src/jaegertracing/opentelemetry-cpp/api/include -isystem /ceph/src/jaegertracing/opentelemetry-cpp/exporters/jaeger/include -isystem /ceph/src/jaegertracing/opentelemetry-cpp/ext/include -isystem /ceph/src/jaegertracing/opentelemetry-cpp/sdk/include -isystem /ceph/src/xxHash -isystem /ceph/src/fmt/include -isystem /ceph/build/src/breakpad_project-prefix/include -isystem /ceph/build/src/breakpad_project-prefix/include/breakpad -isystem /ceph/build/src/pmdk/src/include -g -Werror -fPIC -fno-builtin-malloc -fno-builtin-calloc -fno-builtin-realloc -fno-builtin-free -Wall -fno-strict-aliasing -fsigned-char -Wtype-limits -Wignored-qualifiers -Wpointer-arith -Werror=format-security -Winit-self -Wno-unknown-pragmas -Wnon-virtual-dtor -Wno-ignored-qualifiers -Wsign-compare -ftemplate-depth-1024 -DBOOST_ALLOW_DEPRECATED_HEADERS -Wpessimizing-move -Wredundant-move -Wno-inconsistent-missing-override -Wno-mismatched-tags -Wno-unused-private-field -Wno-address-of-packed-member -Wno-unknown-warning-option -Wno-unused-function -Wno-unused-local-typedef -Wno-varargs -Wno-gnu-designator -Wno-missing-braces -Wno-parentheses -Wno-deprecated-register -Wno-vla-cxx-extension -DCEPH_DEBUG_MUTEX -D_GLIBCXX_ASSERTIONS -fdiagnostics-color=auto -std=c++2b -MD -MT src/librbd/CMakeFiles/rbd_internal.dir/mirror/GroupPromoteRequest.cc.o -MF src/librbd/CMakeFiles/rbd_internal.dir/mirror/GroupPromoteRequest.cc.o.d -o src/librbd/CMakeFiles/rbd_internal.dir/mirror/GroupPromoteRequest.cc.o -c /ceph/src/librbd/mirror/GroupPromoteRequest.cc
/ceph/src/librbd/mirror/GroupPromoteRequest.cc:35:16: error: unused variable 'MAX_RETURN' [-Werror,-Wunused-const-variable]
   35 | const uint32_t MAX_RETURN = 1024;
      |                ^~~~~~~~~~
1 error generated.
```

--- a/src/librbd/mirror/GroupEnableRequest.cc
+++ b/src/librbd/mirror/GroupEnableRequest.cc
@@ -235,7 +235,7 @@ void GroupEnableRequest<I>::create_primary_group_snapshot() {
     lderr(m_cct) << "failed to retrieve min OSD release: " << cpp_strerror(r)
                  << dendl;
     m_ret_val = r;
-    close_images();
+    disable_mirror_group();
     return;
   }
 
@@ -272,7 +272,7 @@ void GroupEnableRequest<I>::handle_create_primary_group_snapshot(int r) {
                  << cpp_strerror(r) << dendl;
     m_ret_val = r;
 
-    close_images();
+    disable_mirror_group();
     return;
   }
 
@@ -325,7 +325,7 @@ void GroupEnableRequest<I>::handle_create_primary_image_snapshots(int r) {
       m_group_snap.snaps[i].snap_id = m_snap_ids[i];
     }
 
-    remove_primary_group_snapshot();
+    disable_mirror_group();
     return;
   }
 
@@ -364,7 +364,7 @@ void GroupEnableRequest<I>::handle_update_primary_group_snapshot(int r) {
                  << cpp_strerror(r) << dendl;
     m_ret_val = r;
 
-    remove_primary_group_snapshot();
+    disable_mirror_group();
     return;
   }
 

--- a/src/librbd/mirror/GroupPromoteRequest.cc
+++ b/src/librbd/mirror/GroupPromoteRequest.cc
@@ -30,12 +30,6 @@
 namespace librbd {
 namespace mirror {
 
-namespace {
-
-const uint32_t MAX_RETURN = 1024;
-
-} // anonymous namespace
-
 using util::create_context_callback;
 using util::create_rados_callback;
 
@@ -533,58 +527,48 @@ void GroupPromoteRequest<I>::handle_acquire_exclusive_locks(int r) {
   if (!m_need_rollback) {
     ldout(m_cct, 10) << "no rollback required" << dendl;
     create_primary_group_snapshot();
-  } else {
-    // current membership
-    std::vector<cls::rbd::GroupImageSpec> current_membership;
-    for (const auto& image : m_images) {
-      current_membership.push_back(image.spec);
-    }
-
-    // rollback membership
-    auto* rollback_snap = &m_rollback_group_snap;
-    std::vector<cls::rbd::GroupImageSpec> rollback_membership;
-    for (auto& it : rollback_snap->snaps) {
-      rollback_membership.emplace_back(it.image_id, it.pool);
-    }
-
-    if (rollback_membership != current_membership) {
-      ldout(m_cct, 10) << "rollback group snapshot membership with snap id: "
-                       << rollback_snap->id
-                       << ", does not match current group membership"
-                       << dendl;
-      fix_group_membership(current_membership, rollback_membership);
-      return;
-    }
-    rollback();
+    return;
   }
-}
 
-template <typename I>
-void GroupPromoteRequest<I>::fix_group_membership(
-    std::vector<cls::rbd::GroupImageSpec>& current_membership,
-    std::vector<cls::rbd::GroupImageSpec>& rollback_membership) {
-  ldout(m_cct, 10) << dendl;
+  // current membership
+  std::vector<cls::rbd::GroupImageSpec> current_membership;
+  for (const auto& image : m_images) {
+    current_membership.push_back(image.spec);
+  }
 
-  m_to_add.clear();
-  m_to_remove.clear();
+  // rollback membership
+  auto* rollback_snap = &m_rollback_group_snap;
+  std::vector<cls::rbd::GroupImageSpec> rollback_membership;
+  for (auto& it : rollback_snap->snaps) {
+    rollback_membership.emplace_back(it.image_id, it.pool);
+  }
+
+  if (rollback_membership == current_membership) {
+    rollback();
+    return;
+  }
+
+  ldout(m_cct, 10) << "rollback group snapshot membership with snap id: "
+                   << rollback_snap->id
+                   << ", does not match current group membership"
+                   << dendl;
 
   std::set<std::pair<std::string, int64_t>> current_set;
   std::set<std::pair<std::string, int64_t>> rollback_set;
-
   for (auto& img : current_membership) {
     current_set.insert({img.image_id, img.pool_id});
   }
-
   for (auto& img : rollback_membership) {
     rollback_set.insert({img.image_id, img.pool_id});
   }
 
+  m_to_add.clear();
+  m_to_remove.clear();
   for (auto& img : rollback_membership) {
     if (!current_set.count({img.image_id, img.pool_id})) {
       m_to_add.push_back(img);
     }
   }
-
   for (auto& img : current_membership) {
     if (!rollback_set.count({img.image_id, img.pool_id})) {
       m_to_remove.push_back(img);
@@ -594,7 +578,6 @@ void GroupPromoteRequest<I>::fix_group_membership(
   ldout(m_cct, 10) << "fixing group membership, "
                    << "to_add=" << m_to_add.size()
                    << ", to_remove=" << m_to_remove.size() << dendl;
-
   if (!m_to_add.empty()) {
     lderr(m_cct) << "rollback requires adding images to group, but dynamic "
                  << "group removal is not supported today" << dendl;
@@ -602,8 +585,7 @@ void GroupPromoteRequest<I>::fix_group_membership(
     release_exclusive_locks();
     return;
   }
-
-  if (!m_to_remove.empty() && m_to_remove.size() > 1) {
+  if (m_to_remove.size() > 1) {
     lderr(m_cct) << "rollback requires more than one image to be removed from "
                  << "the group, this is not supported today" << dendl;
     m_ret_val = -EINVAL;
@@ -667,54 +649,22 @@ void GroupPromoteRequest<I>::handle_remove_images_from_group(int r) {
     }
   }
 
-  m_images.clear();
-  list_group_images();
-}
+  // Remove from m_images
+  for (auto& spec : m_to_remove) {
+    auto it = std::remove_if(
+        m_images.begin(), m_images.end(),
+        [&](const cls::rbd::GroupImageStatus& image) {
+          return image.spec.image_id == spec.image_id &&
+                 image.spec.pool_id == spec.pool_id;
+        });
 
-template <typename I>
-void GroupPromoteRequest<I>::list_group_images() {
-  ldout(m_cct, 10) << dendl;
-
-  librados::ObjectReadOperation op;
-  cls_client::group_image_list_start(&op, m_start_after, MAX_RETURN);
-
-  auto comp = create_rados_callback<
-    GroupPromoteRequest<I>,
-    &GroupPromoteRequest<I>::handle_list_group_images>(this);
-
-  m_out_bl.clear();
-  int r = m_group_ioctx.aio_operate(
-    librbd::util::group_header_name(m_group_id), comp, &op, &m_out_bl);
-  ceph_assert(r == 0);
-  comp->release();
-}
-
-template <typename I>
-void GroupPromoteRequest<I>::handle_list_group_images(int r) {
-  ldout(m_cct, 10) << "r=" << r << dendl;
-
-  std::vector<cls::rbd::GroupImageStatus> images;
-  if (r == 0) {
-    auto iter = m_out_bl.cbegin();
-    r = cls_client::group_image_list_finish(&iter, &images);
+    if (it != m_images.end()) {
+      m_images.erase(it, m_images.end());
+    }
   }
 
-  if (r < 0) {
-    lderr(m_cct) << "group_id=" << m_group_id
-                 << " error listing images in group: "
-                 << cpp_strerror(r) << dendl;
-    m_ret_val = r;
-    release_exclusive_locks();
-    return;
-  }
-
-  auto image_count = images.size();
-  m_images.insert(m_images.end(), images.begin(), images.end());
-  if (image_count == MAX_RETURN) {
-    m_start_after = images.rbegin()->spec;
-    list_group_images();
-    return;
- }
+  ldout(m_cct, 10) << "updated membership: "
+                   << "remaining images=" << m_images.size() << dendl;
 
   rollback();
 }

--- a/src/librbd/mirror/GroupPromoteRequest.h
+++ b/src/librbd/mirror/GroupPromoteRequest.h
@@ -83,13 +83,7 @@ private:
    * ACQUIRE_EXCLUSIVE_LOCKS ---------------------------|
    *    |                                               |
    *    v  (skip if not needed)                         |
-   * FIX_GROUP_MEMBERSHIP                               |
-   *    |                                               |
-   *    v  (skip if not needed)                         |
    * REMOVE_IMAGES_FROM_GROUP                           |
-   *    |                                               |
-   *    v  (skip if not needed)                         |
-   * LIST_GROUP_IMAGES                                  |
    *    |                                               |
    *    v                                               |
    * ROLLBACK                                           |
@@ -197,15 +191,8 @@ private:
   void acquire_exclusive_locks();
   void handle_acquire_exclusive_locks(int r);
 
-  void fix_group_membership(
-      std::vector<cls::rbd::GroupImageSpec>& current_membership,
-      std::vector<cls::rbd::GroupImageSpec>& rollback_membership);
-
   void remove_images_from_group();
   void handle_remove_images_from_group(int r);
-
-  void list_group_images();
-  void handle_list_group_images(int r);
 
   void rollback();
   void handle_rollback(int r);

